cmd/kube-apiserver/app/server.go

```diff
@@ -381,6 +383,53 @@ func CreateKubeAPIServerConfig(
 		config.ExtraConfig.KubeletClientConfig.Lookup = config.GenericConfig.EgressSelector.Lookup
 	}
 
+	// If ServiceAccountIssuerDiscovery is enabled, construct the OIDC metadata server.
+	if utilfeature.DefaultFeatureGate.Enabled(features.ServiceAccountIssuerDiscovery) {
+		var pubKeys []interface{}
+		for _, f := range s.Authentication.ServiceAccounts.KeyFiles {
+			keys, err := keyutil.PublicKeysFromFile(f)
+			if err != nil {
+				return nil, nil, nil, nil, fmt.Errorf("failed to parse service-account-key-file %q: %v", f, err)
+			}
+			pubKeys = append(pubKeys, keys...)
+		}
+
+		// Either use the provided JWKS URI or default to ExternalHost plus
+		// the JWKS path.
+		jwksURI := s.Authentication.ServiceAccounts.JWKSURI
+		if jwksURI == "" {
+			u, err := url.Parse(s.GenericServerRunOptions.ExternalHost)
```

Comment:
should we be using genericConfig.ExternalAddress instead of ExternalHost? That is
* based on ExternalHost: https://github.com/kubernetes/kubernetes/blob/70551475854a8854cb5feb8762a9a0a5e39900a5/staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go#L73-L76
* adds in the secure port we are listening on if ExternalHost has no port: https://github.com/kubernetes/kubernetes/blob/70551475854a8854cb5feb8762a9a0a5e39900a5/staging/src/k8s.io/apiserver/pkg/server/config.go#L428-L438
* is what is used when serving addresses in discovery docs already: https://github.com/kubernetes/kubernetes/blob/70551475854a8854cb5feb8762a9a0a5e39900a5/pkg/master/master.go#L300

<!--
ExternalHost is populated using the first of:
* explicit `--external-hostname`
* fallback to AdvertiseAddress: https://github.com/kubernetes/kubernetes/blob/70551475854a8854cb5feb8762a9a0a5e39900a5/cmd/kube-apiserver/app/server.go#L586-L588
* fallback to os.Hostname():
https://github.com/kubernetes/kubernetes/blob/70551475854a8854cb5feb8762a9a0a5e39900a5/cmd/kube-apiserver/app/server.go#L589-L591

AdvertiseAddress is populated using the first of:
* explicit `--advertise-address`
* fallback to `--bind-address`:
https://github.com/kubernetes/kubernetes/blob/70551475854a8854cb5feb8762a9a0a5e39900a5/pkg/kubeapiserver/options/serving.go#L62-L68
* fallback to the host's default IP:
https://github.com/kubernetes/kubernetes/blob/70551475854a8854cb5feb8762a9a0a5e39900a5/staging/src/k8s.io/apimachinery/pkg/util/net/interface.go#L431-L438
-->

Reply:
I'm wondering if, instead, we should construct this from the issuer URL by default?

Reply:
We don't know that the issuer points at the kube-apiserver, and we don't know what subpath (if any) the jwks doc is hosted at. I think we should use ExternalAddress as the default, overrideable using JWKSURI

Reply:
Yeah good point. I think ExternalAddress is probably the best for now then...  good user-facing docs for this are going to be _**SO**_ important.

Reply:
Since the generic config isn't completed until `master.go`, I'll pass the URLs and key file paths through and do the construction of both responses in `master.go` then.

Reply:
Can we expect that `ExternalAddress` is always at most host + port (no scheme)? If so we don't have to try to `url.Parse` it, we can just construct the URL around `ExternalAddress`.


`url.Parse` errors if there's a port but no scheme. I'm assuming because the above construction is using `SplitHostPort` and then `JoinHostPort` that `ExternalAddress` is expected to be at most host + port, since having a scheme _and_ a port breaks `SplitHostPort` (and if it's a scheme but no port it just does the wrong thing, and treats the scheme as the host).

Reply:
Refactored to use `ExternalAddress`. Kept it in a separate commit for now to make it easy to see the changes - a lot moves to `master.go` and instead of trying to parse the address as a URL, we construct a URL from the address and then parse it as a sanity check that `ExternalAddress` met my expectation (just host + port).